keras/layers/convolutional.py

```diff
@@ -2310,7 +2310,75 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 
-class Cropping1D(Layer):
+class _Cropping(Layer):
+    """Abstract nD copping layer (private, used as implementation base).
+
+    # Arguments
+        rank: An integer, the rank of the cropping,
+            e.g. "2" for Cropping2D.
+        cropping: A tuple of tuples of 2 ints. For Cropping1D,
+            it is just a tuple of two ints.
+        data_format: A string,
+            one of `"channels_last"` or `"channels_first"`.
+            The ordering of the dimensions in the inputs.
+            `"channels_last"` corresponds to inputs with shape
+            `(batch, ..., channels)` while `"channels_first"` corresponds to
+            inputs with shape `(batch, channels, ...)`.
+            It defaults to the `image_data_format` value found in your
+            Keras config file at `~/.keras/keras.json`.
+            If you never set it, then it will be "channels_last".
+            For Cropping1D, the data format is always `"channels_last"`.
+    """
+
+    def __init__(self, rank,
+                 cropping,
+                 data_format=None,
+                 **kwargs):
+        super(_Cropping, self).__init__(**kwargs)
+        self.rank = rank
+        self.cropping = cropping
+        self.data_format = K.normalize_data_format(data_format)
+        self.input_spec = InputSpec(ndim=2 + rank)
+
+    def call(self, inputs):
+        slices_dims = []
+        for start, end in self._normalized_cropping:
+            if end == 0:
+                end = None
+            else:
+                end = -end
+            slices_dims.append(slice(start, end))
+
+        slices = [slice(None)] + slices_dims + [slice(None)]
+        slices = tuple(slices)
+        spatial_axes = list(range(1, 1 + self.rank))
+        slices = transpose_shape(slices, self.data_format, spatial_axes)
+        return inputs[slices]
+
+    def compute_output_shape(self, input_shape):
+        cropping_all_dims = ((0, 0),) + self._normalized_cropping + ((0, 0),)
+        spatial_axes = list(range(1, 1 + self.rank))
+        cropping_all_dims = transpose_shape(cropping_all_dims,
+                                            self.data_format,
+                                            spatial_axes)
+        output_shape = list(input_shape)
+        for dim in range(len(output_shape)):
+            if output_shape[dim] is not None:
+                output_shape[dim] -= sum(cropping_all_dims[dim])
+        return tuple(output_shape)
+
+    def get_config(self):
+        config = {'cropping': self.cropping,
+                  'data_format': self.data_format}
+        base_config = super(_Cropping, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    @property
+    def _normalized_cropping(self):
```

Comment:
This property is overwridden in Cropping1D, that is why I can't use just `self.cropping`, it is not always a tuple of tuples.

Reply:
This overriding mechanism seems complicated. Especially since in the code you refer to the `cropping` argument in `Cropping1D` as `normalized_cropping `, although `self._normalized_cropping` returns a *different* value...

Why would you need more than a single `cropping` argument? Make the base class normalize the cropping argument to be a tuple of tuples of 2 ints, of the same length as the rank.

Reply:
Ok sure thing, thanks, since `self.cropping` is public, I wanted to avoid changing it, but what you are describing will make the code easier to read.